clab/dependency_manager/dependency_node.go

```diff
@@ -0,0 +1,118 @@
+package dependency_manager
+
+import (
+	"fmt"
+	"sync"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/srl-labs/containerlab/types"
+)
+
+// dependencyNode is the representation of a node in the dependency concept.
+type dependencyNode struct {
+	name                string
+	stageBarrier        map[types.WaitForStage]*sync.WaitGroup
+	stageBarrierCounter map[types.WaitForStage]uint
+	depender            map[types.WaitForStage][]*dependerNodeStage
+
+	m sync.Mutex
+}
+
+// newDependencyNode initializes a dependencyNode with the given name.
+func newDependencyNode(name string) *dependencyNode {
+	d := &dependencyNode{
+		name:                name,
+		stageBarrier:        map[types.WaitForStage]*sync.WaitGroup{},
+		stageBarrierCounter: map[types.WaitForStage]uint{},
+		depender:            map[types.WaitForStage][]*dependerNodeStage{},
+	}
```

Comment:
Do we need `stageBarrierCounter` at all? Is it different from `len(d.depender[stage])`?

Reply:
Changed stageBarrierCounter to MustWait, which is only a boolean value, indicating if the given stage is to be waited for or it can be continued with the next phase.